apps/gradient_autoscheduler/GradientAutoscheduler.cpp

```diff
@@ -0,0 +1,1008 @@
+        schedule_source << "    .vectorize(" <<
+            vectorized_rvar.name() << ")\n";
+    }
+}
+
+template <typename FuncOrStage>
+void parallelize_vars_and_rvars(
+        const MachineParams &params,
+        FuncOrStage func_or_stage,
+        int natural_vector_size,
+        bool is_pure_def,
+        const std::vector<Var> &vars,
+        const std::vector<int> &var_bounds,
+        const std::vector<RVar> &rvars,
+        const std::vector<int> &rvar_bounds,
+        TailStrategy tail,
+        bool is_gpu,
+        std::ostringstream &schedule_source) {
+    if (is_gpu) {
+        return parallelize_vars_and_rvars_gpu(
+            params,
+            func_or_stage,
+            is_pure_def,
+            vars,
+            var_bounds,
+            rvars,
+            rvar_bounds,
+            tail,
+            schedule_source);
+    } else {
+        return parallelize_vars_and_rvars_cpu(
+            params,
+            func_or_stage,
+            natural_vector_size,
+            is_pure_def,
+            vars,
+            var_bounds,
+            rvars,
+            rvar_bounds,
+            tail,
+            schedule_source);
+    }
+}
+
+void apply_schedule(const MachineParams &params,
+                    const Target &target,
+                    Func func,
+                    int update_id,
+                    const std::vector<int> &var_bounds,
+                    bool is_gpu,
+                    std::ostringstream &schedule_source) {
+    if (update_id == -1) {
+        func.compute_root();
+        schedule_source << func.name() << ".compute_root()\n";
+        if (func.dimensions() > 0) {
+            parallelize_vars_and_rvars(
+                params,
+                func,
+                natural_vector_size(target, func.values()[0].type()),
+                true, // is_pure_def
+                func.args(),
+                var_bounds,
+                {}, // rvars
+                {}, // rvar_bounds
+                TailStrategy::ShiftInwards,
+                is_gpu,
+                schedule_source);
+        }
+    } else {
+        // If the pure domain is smaller than the reduction domain,
+        // we try to apply rfactor to increase parallelism:
+        bool checked_associative = false;
+        bool is_associative = false;
+        int domain_size = 1;
+        for (int b : var_bounds) {
+            domain_size *= b;
+        }
+        std::vector<ReductionVariable> reduction_vars =
+            func.update(update_id).get_schedule().rvars();
+        std::vector<int> rvar_bounds = get_rvar_bounds(reduction_vars);
+        std::vector<RVar> rvars;
+        rvars.reserve(reduction_vars.size());
+        for (ReductionVariable r : reduction_vars) {
+            rvars.push_back(RVar(r.var));
+        }
+        int rdomain_size = 1;
+        for (int b : rvar_bounds) {
+            rdomain_size *= b;
+        }
+        int cpu_max_domain_size = 8 * params.parallelism;
```

Comment:
Please add comments about why these constants (8, 4096, etc) were chosen.

Reply:
Done. In general these numbers can be improved.